pkg/apiserver/controllers/tunnel_agent_reconciler.go

```diff
@@ -0,0 +1,316 @@
+	"github.com/apoxy-dev/apoxy/pkg/tunnel/vni"
+)
+
+const (
+	indexByTunnelRef = "spec.tunnelRef.name"
+)
+
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnelagents,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnelagents/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnelagents/finalizers,verbs=update
+// +kubebuilder:rbac:groups=core.apoxy.dev/v1alpha2,resources=tunnels,verbs=get;list;watch
+
+type TunnelAgentReconciler struct {
+	client    client.Client
+	agentIPAM tunnet.IPAM
+	vniPool   *vni.VNIPool
+}
+
+func NewTunnelAgentReconciler(c client.Client, agentIPAM tunnet.IPAM, vniPool *vni.VNIPool) *TunnelAgentReconciler {
+	return &TunnelAgentReconciler{
+		client:    c,
+		agentIPAM: agentIPAM,
+		vniPool:   vniPool,
+	}
+}
+
+func (r *TunnelAgentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := controllerlog.FromContext(ctx, "name", req.Name)
+
+	var agent corev1alpha2.TunnelAgent
+	if err := r.client.Get(ctx, req.NamespacedName, &agent); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// handle deletion
+	if !agent.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(&agent, ApiServerFinalizer) {
+			if err := r.releaseResourcesIfPresent(log, &agent); err != nil {
+				log.Error(err, "failed to release resources; will retry")
+				return ctrl.Result{}, fmt.Errorf("failed to release resources: %w", err)
+			}
+
+			// Remove finalizer
+			if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+				var cur corev1alpha2.TunnelAgent
+				if getErr := r.client.Get(ctx, req.NamespacedName, &cur); getErr != nil {
+					return getErr
+				}
+				controllerutil.RemoveFinalizer(&cur, ApiServerFinalizer)
+				return r.client.Update(ctx, &cur)
+			}); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// ensure finalizer
+	if !controllerutil.ContainsFinalizer(&agent, ApiServerFinalizer) {
+		if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			var cur corev1alpha2.TunnelAgent
+			if getErr := r.client.Get(ctx, req.NamespacedName, &cur); getErr != nil {
+				return getErr
+			}
+			controllerutil.AddFinalizer(&cur, ApiServerFinalizer)
+			return r.client.Update(ctx, &cur)
+		}); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// fetch owner Tunnel
+	tunnelName := agent.Spec.TunnelRef.Name
+	if tunnelName == "" {
+		log.Info("tunnelRef.name is empty; skipping")
+		return ctrl.Result{}, nil
+	}
+
+	var tunnel corev1alpha2.Tunnel
+	if err := r.client.Get(ctx, client.ObjectKey{Name: tunnelName}, &tunnel); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Referenced Tunnel not found; will retry", "tunnel", tunnelName)
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
```

Comment:
Don't think it's ever going to help - if Tunnel doesn't exist already something is very wrong. Can probably set status to "Failed" or something